### PR TITLE
New strategy: mutually broadcastable shapes

### LINKED ISSUE #1969: Add Numpy strategy for `multiple_shapes` which are mutually broadcastable
The `broadcastable_shapes(shape)` strategy is great, but sometimes you need *several* shapes, all of which are broadcastable against any of the others - for example to describe the [loop dimensions of generalised ufuncs](https://docs.scipy.org/doc/numpy/reference/c-api.generalized-ufuncs.html).  This may not be needed often, but be critical when it is!

I therefore propose to add the following strategy:

```python
Shape = Tuple[int, ...]

def multiple_shapes(
    inputs,
    base_shape=(),
    min_dims=0,
    max_dims=None,
    min_side=1,
    max_side=None,
):
    # type: (int, Shape, int, int, int, int) -> Tuple[Tuple[Shape, ...], Shape]
    ...
    # Returns a namedtuple of (input_shapes, result_shape)
```

This strategy will be defined in the first release as generating an `len(inputs)`-tuple of mutually-broadcastable array shapes, and the expected result shape as the second element of a named tuple.

In future releases, `inputs` may be a gufunc signature string describing core dimensions, with the other arguments interpreted as describing the loop dimensions (if any).  This plan replaces #1784.

### PR DESCRIPTION
**Pull request by rsokl**
Closes #1969 and replaces the mammoth #2153

### CODE REVIEW

**Comment by Zac-HD** [comment]
We'll get there eventually!  Thanks for you persistence Ryan :sparkling_heart: 

Note to self: once this is merged,

- [ ] Open an issue to use strict validation of shape/dims/side compatibility in `broadcastable_shapes` (with deprecation pathway)
- [ ] Open an issue to support a `ufunc=` signature-or-function argument to `mutually_broadcastable_shapes`, as an alternative to `num_shapes`

(not opening yet because the details will depend on what we do here, and not including in this PR because they're all distinct user-facing changes and the PR is already huge.)

File: hypothesis-python/tests/numpy/test_gen_data.py [hunk 1]
```diff
@@ -655,10 +700,25 @@ def test_broadcastable_shape_has_good_default_values(shape, data):
...
+    assert len(shapes) == num_shapes
+    # raises if shapes are not mutually-compatible
+    assert result == _broadcast_shapes(base_shape, *shapes)
+
+
 @settings(deadline=None)
 @given(min_dim=st.integers(0, 32), shape=ANY_SHAPE, data=st.data())
 def test_broadcastable_shape_can_broadcast(min_dim, shape, data):
-    max_dim = data.draw(st.none() | st.integers(min_dim, 32), label="max_dim")
+    max_dim = data.draw(st.none() | st.integers(min_dim, 32), label="max_dims")
```
Comment (Zac-HD) [inline hypothesis-python/tests/numpy/test_gen_data.py]: Here, and in the other test, you may want to rename `min_dim -> min_dims` too for consistency.

**Comment by rsokl** [comment]
@Zac-HD the test that is failing has popped up a couple of times now. Seems like another flaky test.

**Comment by Zac-HD** [comment]
🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉 🎉

**Pull request merged**